src/Compilers/Core/Portable/Operations/CustomDataFlowAnalysis.cs

```diff
@@ -0,0 +1,402 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis
+{
+#if WORKSPACE
+    using TBasicBlock = BasicBlock;
+    using TControlFlowBranch = ControlFlowBranch;
+#else
+    using TBasicBlock = BasicBlockBuilder;
+    using TControlFlowBranch = BasicBlockBuilder.Branch;
+#endif
+
+    internal static class CustomDataFlowAnalysis<TDataFlowAnalyzer, TBlockAnalysisData>
+        where TDataFlowAnalyzer: IDataFlowAnalyzer<TBlockAnalysisData>
+    {
+        /// <summary>
+        /// Runs dataflow analysis for the given <paramref name="analyzer"/> on the given <paramref name="blocks"/>.
+        /// </summary>
+        /// <param name="blocks">Blocks on which to execute analysis.</param>
+        /// <param name="analyzer">Dataflow analyzer.</param>
+        /// <returns>Block analysis data for the last block.</returns>
+        internal static TBlockAnalysisData Run(ImmutableArray<TBasicBlock> blocks, TDataFlowAnalyzer analyzer, CancellationToken cancellationToken)
```

Comment:
This is mostly a port of existing basic block reachability analysis from the ControlFlowGraphBuilder type.